### PR TITLE
ci: Add annotations, more mypy checks

### PR DESCRIPTION
**Pull request by mdeweerd**
- Improve caching by also saving cache on failures.
- Generate github annotations from pre-commit output (need logToCs.py);
- Install typing dependencies to remove "ignore-missing-imports" on mypy
- Add mdformat plugins to format code.
- Ignore test on cygwin
- Add codespell checks

Note: regarding the annotations, you can have a look in the [Files changed tab](https://github.com/Lucas-C/pre-commit-hooks/pull/95/files#file-pre_commit_hooks-insert_license-py-L613) .
The action summary also has up to 10 error and 10 warning annotations.

### CODE REVIEW

File: tests/chmod_test.py [hunk 1]
```diff
@@ -12,7 +12,11 @@ def test_chmod_ok():
             assert (
                 "This hook does nothing when executed on Windows" in stdout.getvalue()
             )
+        elif sys.platform == "cygwin":
```
Comment (Lucas-C) [inline tests/chmod_test.py:15]: Why this change?

I used to work on Cygwin, and never had any issue with this hook. The `chmod()` function worked fine.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -3,8 +3,6 @@ name: CI
 
 on:  # cf. https://github.community/t/how-to-trigger-an-action-on-push-or-pull-request-but-not-both/16662
   push:
-    branches:
-      - master
```
Comment (Lucas-C) [inline .github/workflows/ci.yml:7]: Why this change?

Isn't it better to only execute this CI pipeline on PRs targetting the `master` branch?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -47,5 +49,40 @@ jobs:
           pip install --upgrade .
 
       - name: Running checks ☑
+        id: precommit
         if: matrix.python-version != '3.11'  # Pylint currently triggers erroneous no-member errors
-        run: pre-commit run --all-files --verbose
+        uses: mathiasvr/command-output@v2.0.0
```
Comment (Lucas-C) [inline .github/workflows/ci.yml:54]: Why introducing a dependency to another GA workflow script?

Unless it's necesseray or really useful, I'd prefer to limit the number of dependencies.

Having many such dependencies statistically increases the risk of of them breaking in the future, causing some failures & maintenance costs (time to identify & fix the problem) to this project.

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Overall, I really like the idea of `logToCs.py` to convert `pre-commit` output into a CheckStyle report, so that they can then be exposed as PR annotations.
Good job on cradfting this 👍

However, I am NOT favourable towards adding this ~400 lines Python script to this project.
It's purpose is distinct / larger than the scope of this repository, and I do not wish to maintain it.

It would be better as a standalone project! 🙂
If published on Pypi, or in a dedicated GitHub repo usable as a GA workflow script, it could be reused in several GitHub Actions pipelines! And it could have its own `README` with explanations and even potentially unit tests.

Again, to be clear: I would gladly consider using this tool if it only introduced a few extra lines in `.github/workflows/ci.yml`, but I'm sorry I won't merge a PR that introduces hundreds of new lines of Python code that are not directly related to `pre-commit` hooks.

File: tests/chmod_test.py (hunk 1 above)
Comment (mdeweerd) [reply tests/chmod_test.py:15]: I appreciate your interest in Cygwin - most of the time I'll get a "Cygwin is not supported" and that has been 100% the case up-to now for pre-commit related projects even if I propose a fix.

On my system the 'x' bits are set according to the contents of the file - a file with a shebang will show up executable.

So my guess is that it's related to that behavior, so I dig not into it much deeper and decided to disable that test on Cygwin to get past a failing pre-commit run.

I do not use it, but I think it sufficiently works for one the purpose of it: setting executable bits for the git repository.

**Comment by mdeweerd** [comment]
I am not at the step of setting up a package or GA yet - I have set up a project for it though:
https://github.com/mdeweerd/LogToCheckStyle

**Comment by mdeweerd** [comment]
FYI: https://github.com/mdeweerd/LogToCheckStyle now proposes a Github Action.
